HPX can't compile with GCC 7.3 on x86_64

## Expected Behavior

HPX compiles

## Actual Behavior

HPX fails to compile.  The compiler error is:

```
In file included from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:7:0:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads/executors/current_executor.hpp:95:12 error: using typedef-name ‘using current_executor = class hpx::parallel::execution::thread_pool_executor’ after ‘struct’
     struct current_executor : public scheduled_executor
            ^~~~~~~~~~~~~~~~
In file included from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads/executors/current_executor.hpp:15:0,
                 from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:7:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads_fwd.hpp:26:79: note: ‘using current_executor = class hpx::parallel::execution::thread_pool_executor’ has a previous declaration here
             using current_executor = parallel::execution::thread_pool_executor;
                                                                               ^
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp: In member function ‘hpx::state hpx::threads::executors::detail::current_executor::get_state() const’:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:150:30: error: ‘get_thread_num_tss’ is not a member of ‘hpx::threads::detail’
             threads::detail::get_thread_num_tss());
                              ^~~~~~~~~~~~~~~~~~
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:150:30: note: suggested alternative: ‘get_thread_nums_tss’
             threads::detail::get_thread_num_tss());
                              ^~~~~~~~~~~~~~~~~~
                              get_thread_nums_tss
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp: At global scope:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:165:40: error: ISO C++ forbids declaration of ‘current_executor’ with no type [-fpermissive]
     current_executor::current_executor()
                                        ^
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:165:40: error: invalid use of incomplete type ‘class hpx::parallel::execution::thread_pool_executor’
In file included from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads/executors/current_executor.hpp:15:0,
                 from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:7:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads_fwd.hpp:19:26: note: forward declaration of ‘class hpx::parallel::execution::thread_pool_executor’
         class HPX_EXPORT thread_pool_executor;
                          ^~~~~~~~~~~~~~~~~~~~
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:170:39: error: expected constructor, destructor, or type conversion before ‘(’ token
     current_executor::current_executor(policies::scheduler_base* scheduler)
                                       ^
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:174:46: error: invalid use of incomplete type ‘class hpx::parallel::execution::thread_pool_executor’
     hpx::state current_executor::get_state() const
                                              ^~~~~
In file included from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads/executors/current_executor.hpp:15:0,
                 from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:7:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads_fwd.hpp:19:26: note: forward declaration of ‘class hpx::parallel::execution::thread_pool_executor’
         class HPX_EXPORT thread_pool_executor;
                          ^~~~~~~~~~~~~~~~~~~~
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp: In function ‘hpx::threads::executors::current_executor hpx::threads::get_executor(const thread_id_type&, hpx::error_code&)’:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:184:49: error: return type ‘using current_executor = class hpx::parallel::execution::thread_pool_executor {aka class hpx::parallel::execution::thread_pool_executor}’ is incomplete
         thread_id_type const& id, error_code& ec)
                                                 ^
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:183:33: error: ambiguating new declaration of ‘void hpx::threads::get_executor(const thread_id_type&, hpx::error_code&)’
     executors::current_executor get_executor(
                                 ^~~~~~~~~~~~
In file included from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:7:0:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads/executors/current_executor.hpp:118:57: note: old declaration ‘hpx::threads::executors::current_executor hpx::threads::get_executor(const thread_id_type&, hpx::error_code&)’
     HPX_API_EXPORT threads::executors::current_executor get_executor(
                                                         ^~~~~~~~~~~~
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:190:55: error: invalid use of incomplete type ‘using current_executor = class hpx::parallel::execution::thread_pool_executor {aka class hpx::parallel::execution::thread_pool_executor}’
             return executors::current_executor(nullptr);
                                                       ^
In file included from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads/executors/current_executor.hpp:15:0,
                 from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:7:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads_fwd.hpp:19:26: note: forward declaration of ‘using current_executor = class hpx::parallel::execution::thread_pool_executor {aka class hpx::parallel::execution::thread_pool_executor}’
         class HPX_EXPORT thread_pool_executor;
                          ^~~~~~~~~~~~~~~~~~~~
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:197:57: error: invalid use of incomplete type ‘using current_executor = class hpx::parallel::execution::thread_pool_executor {aka class hpx::parallel::execution::thread_pool_executor}’
             get_thread_id_data(id)->get_scheduler_base());
                                                         ^
In file included from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads/executors/current_executor.hpp:15:0,
                 from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:7:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads_fwd.hpp:19:26: note: forward declaration of ‘using current_executor = class hpx::parallel::execution::thread_pool_executor {aka class hpx::parallel::execution::thread_pool_executor}’
         class HPX_EXPORT thread_pool_executor;
                          ^~~~~~~~~~~~~~~~~~~~
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp: In function ‘hpx::threads::executors::current_executor hpx::this_thread::get_executor(hpx::error_code&)’:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:202:69: error: return type ‘using current_executor = class hpx::parallel::execution::thread_pool_executor {aka class hpx::parallel::execution::thread_pool_executor}’ is incomplete
     threads::executors::current_executor get_executor(error_code& ec)
                                                                     ^
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:202:42: error: ambiguating new declaration of ‘void hpx::this_thread::get_executor(hpx::error_code&)’
     threads::executors::current_executor get_executor(error_code& ec)
                                          ^~~~~~~~~~~~
In file included from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:7:0:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads/executors/current_executor.hpp:136:53: note: old declaration ‘hpx::threads::executors::current_executor hpx::this_thread::get_executor(hpx::error_code&)’
     HPX_EXPORT threads::executors::current_executor get_executor(
                                                     ^~~~~~~~~~~~
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:204:64: error: invalid use of incomplete type ‘using current_executor = class hpx::parallel::execution::thread_pool_executor {aka class hpx::parallel::execution::thread_pool_executor}’
         return threads::get_executor(threads::get_self_id(), ec);
                                                                ^
In file included from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads/executors/current_executor.hpp:15:0,
                 from /home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/current_executor.cpp:7:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/hpx/runtime/threads_fwd.hpp:19:26: note: forward declaration of ‘using current_executor = class hpx::parallel::execution::thread_pool_executor {aka class hpx::parallel::execution::thread_pool_executor}’
         class HPX_EXPORT thread_pool_executor;
                          ^~~~~~~~~~~~~~~~~~~~
[ 82%] Building CXX object src/CMakeFiles/hpx.dir/runtime/threads/threadmanager_counters.cpp.o
make[2]: *** [src/CMakeFiles/hpx.dir/runtime/threads/executors/current_executor.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/this_thread_executors.cpp: In member function ‘void hpx::threads::executors::detail::this_thread_executor<Scheduler>::run()’:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/this_thread_executors.cpp:352:48: error: ‘struct hpx::threads::detail::reset_tss_helper’ has no member named ‘previous_thread_num’; did you mean ‘previous_global_thread_num’?
             parent_thread_num_ = reset_on_exit.previous_thread_num();
                                                ^~~~~~~~~~~~~~~~~~~
                                                previous_global_thread_num
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/this_thread_executors.cpp: In member function ‘void hpx::threads::executors::detail::this_thread_executor<Scheduler>::add_processing_unit(std::size_t, std::size_t, hpx::error_code&)’:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/this_thread_executors.cpp:444:45: error: ‘get_thread_num_tss’ is not a member of ‘hpx::threads::detail’
         orig_thread_num_ = threads::detail::get_thread_num_tss();
                                             ^~~~~~~~~~~~~~~~~~
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/this_thread_executors.cpp:444:45: note: suggested alternative: ‘get_thread_nums_tss’
         orig_thread_num_ = threads::detail::get_thread_num_tss();
                                             ^~~~~~~~~~~~~~~~~~
                                             get_thread_nums_tss
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/thread_pool_attached_executors.cpp: In member function ‘void hpx::threads::executors::detail::thread_pool_attached_executor<Scheduler>::add(hpx::threads::detail::executor_base::closure_type&&, const hpx::util::thread_description&, hpx::threads::thread_state_enum, bool, hpx::threads::thread_stacksize, hpx::threads::thread_schedule_hint, hpx::error_code&)’:
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/thread_pool_attached_executors.cpp:78:56: error: ‘get_thread_num_tss’ is not a member of ‘hpx::threads::detail’
             static_cast<std::int16_t>(threads::detail::get_thread_num_tss());
                                                        ^~~~~~~~~~~~~~~~~~
/home/users/khuck/src/phylanx/tools/buildbot/src/hpx/src/runtime/threads/executors/thread_pool_attached_executors.cpp:78:56: note: suggested alternative: ‘get_thread_nums_tss’
             static_cast<std::int16_t>(threads::detail::get_thread_num_tss());
                                                        ^~~~~~~~~~~~~~~~~~
                                                        get_thread_nums_tss
make[2]: *** [src/CMakeFiles/hpx.dir/runtime/threads/executors/thread_pool_attached_executors.cpp.o] Error 1
make[2]: *** [src/CMakeFiles/hpx.dir/runtime/threads/executors/this_thread_executors.cpp.o] Error 1
make[1]: *** [src/CMakeFiles/hpx.dir/all] Error 2
make: *** [all] Error 2
```

## Steps to Reproduce the Problem

... Please be as specific as possible while describing how to reproduce your problem.

  1.  configure with: cmake -DCMAKE_CXX_COMPILER=/packages/gcc/7.3/bin/g++ -DCMAKE_C_COMPILER=/packages/gcc/7.3/bin/gcc -DCMAKE_BUILD_TYPE=Debug -DBOOST_ROOT=/usr/local/packages/boost/1.65.0-gcc7 -DTCMALLOC_ROOT=/usr/local/packages/gperftools/git-gcc7 -DHPX_WITH_MALLOC=tcmalloc -DHWLOC_ROOT= -DCMAKE_INSTALL_PREFIX=. -DHPX_WITH_THREAD_IDLE_RATES=ON -DHPX_WITH_PARCELPORT_MPI=ON -DHPX_WITH_PARCEL_COALESCING=OFF -DHPX_WITH_TOOLS=OFF -DHPX_WITH_TESTS=ON -DHPX_WITH_EXAMPLES=OFF -DHPX_WITH_THREAD_MANAGER_IDLE_BACKOFF=OFF -DHPX_WITH_PARCELPORT_LIBFABRIC=OFF -DHPX_WITH_APEX=TRUE -DAPEX_WITH_ACTIVEHARMONY=TRUE -DACTIVEHARMONY_ROOT=/usr/local/packages/activeharmony/4.6.0-gcc -DAPEX_WITH_OTF2=TRUE -DOTF2_ROOT=/usr/local/packages/otf2-2.1 -DAPEX_WITH_PAPI=TRUE -DPAPI_ROOT=/usr/local/packages/papi/papi-knl/5.5.0/ -DHPX_WITH_APEX_NO_UPDATE=true -DHPX_WITH_APEX_TAG=develop -DHPX_WITH_MAX_CPU_COUNT=72 /home/users/khuck/src/phylanx/tools/buildbot/src/hpx
  1. make
  1. 

## Specifications

... Please describe your environment

  - HPX Version: current master
  - Platform (compiler, OS): gcc 7.3, openmpi 3.1.1, cmake 3.15.1, RHEL 

```
[khuck@delphi buildbot]$ uname -a
Linux delphi 3.10.0-1062.9.1.el7.x86_64 #1 SMP Fri Dec 6 15:49:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

--- a/libs/thread_executors/CMakeLists.txt
+++ b/libs/thread_executors/CMakeLists.txt
@@ -56,7 +56,7 @@ set(thread_executors_sources
 
 include(HPX_AddModule)
 add_hpx_module(thread_executors
-  COMPATIBILITY_HEADERS OFF
+  COMPATIBILITY_HEADERS ON
   DEPRECATION_WARNINGS
   FORCE_LINKING_GEN
   GLOBAL_HEADER_GEN ON

